In [25]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import accuracy_score

# Data collection and analysis

In [4]:
pip install opendatasets --upgrade


In [5]:
import opendatasets as od


In [6]:
dataset_url="https://www.kaggle.com/kumargh/pimaindiansdiabetescsv"
od.download(dataset_url)

Please provide your Kaggle credentials to download this dataset. Learn more: http://bit.ly/kaggle-creds
Your Kaggle username: kushalkumareregowda
Your Kaggle Key: ··········


100%|██████████| 8.89k/8.89k [00:00<00:00, 6.71MB/s]

In [7]:
data_dir='/content/pimaindiansdiabetescsv'

In [8]:
diab_data=pd.read_csv(data_dir+'/pima-indians-diabetes.csv', header=0)

In [9]:
diab_data.head()

6  148  72  35    0  33.6  0.627  50  1
0  1   85  66  29    0  26.6  0.351  31  0
1  8  183  64   0    0  23.3  0.672  32  1
2  1   89  66  23   94  28.1  0.167  21  0
3  0  137  40  35  168  43.1  2.288  33  1
4  5  116  74   0    0  25.6  0.201  30  0

In [10]:
diab_data.shape

(767, 9)

In [11]:
#Statiscal measures
diab_data.describe()

6         148          72  ...       0.627          50           1
count  767.000000  767.000000  767.000000  ...  767.000000  767.000000  767.000000
mean     3.842243  120.859192   69.101695  ...    0.471674   33.219035    0.348110
std      3.370877   31.978468   19.368155  ...    0.331497   11.752296    0.476682
min      0.000000    0.000000    0.000000  ...    0.078000   21.000000    0.000000
25%      1.000000   99.000000   62.000000  ...    0.243500   24.000000    0.000000
50%      3.000000  117.000000   72.000000  ...    0.371000   29.000000    0.000000
75%      6.000000  140.000000   80.000000  ...    0.625000   41.000000    1.000000
max     17.000000  199.000000  122.000000  ...    2.420000   81.000000    1.000000

[8 rows x 9 columns]

#Observations
1. total count in the dataset is 767 records.
2. There are 9 columns in the dataset.


In [12]:
# Diabetes 
diab_data['1'].value_counts()

0    500
1    267
Name: 1, dtype: int64

# Observation
1. we  can see that there are 500 with non diabetic people
2. and 267 people with Diabetic 

In [13]:
diab_data.groupby('1').mean()

6        148         72  ...       33.6     0.627         50
1                                  ...                                
0  3.298000  109.98000  68.184000  ...  30.304200  0.429734  31.190000
1  4.861423  141.23221  70.820225  ...  35.148315  0.550213  37.018727

[2 rows x 8 columns]

In [14]:
# After grouping the data at output level , we uderstood that people who are tested postive ro diabetes are keen to have high levels of
# blodd pressure , insulin , glucose level and so on.

In [15]:
# separating the label from the dataset

In [16]:
x=diab_data.drop(columns='1', axis= 1 ) # as dropping the column #
y=diab_data['1']
print(x)

      6  148  72  35    0  33.6  0.627  50
0     1   85  66  29    0  26.6  0.351  31
1     8  183  64   0    0  23.3  0.672  32
2     1   89  66  23   94  28.1  0.167  21
3     0  137  40  35  168  43.1  2.288  33
4     5  116  74   0    0  25.6  0.201  30
..   ..  ...  ..  ..  ...   ...    ...  ..
762  10  101  76  48  180  32.9  0.171  63
763   2  122  70  27    0  36.8  0.340  27
764   5  121  72  23  112  26.2  0.245  30
765   1  126  60   0    0  30.1  0.349  47
766   1   93  70  31    0  30.4  0.315  23

[767 rows x 8 columns]


In [17]:
print(y)

0      0
1      1
2      0
3      1
4      0
      ..
762    0
763    0
764    0
765    1
766    0
Name: 1, Length: 767, dtype: int64


# Data standaridization 

In [18]:
scaler = StandardScaler()

In [19]:
scaler.fit(x)

StandardScaler()

In [20]:
standardized_data=scaler.transform(x)

In [23]:
X= standardized_data
Y=diab_data['1']

In [22]:
print(X)

[[-0.84372629 -1.12208597 -0.16024856 ... -0.68372895 -0.36426474
  -0.18894038]
 [ 1.23423997  1.94447577 -0.26357823 ... -1.10230105  0.60470064
  -0.1037951 ]
 [-0.84372629 -0.99692019 -0.16024856 ... -0.49346891 -0.91968415
  -1.0403932 ]
 ...
 [ 0.343683    0.0044061   0.14974046 ... -0.73446496 -0.68423462
  -0.27408566]
 [-0.84372629  0.16086333 -0.47023757 ... -0.23978884 -0.37030191
   1.17338414]
 [-0.84372629 -0.8717544   0.04641078 ... -0.20173684 -0.47293375
  -0.87010264]]


# Train and Test data split

In [26]:
x_train, x_test, y_train, y_test = train_test_split(X,Y,test_size=0.2, stratify=Y,random_state=2)

In [30]:
X.shape,x_test.shape , x_train.shape

((767, 8), (154, 8), (613, 8))

In [32]:
classifier=svm.SVC(kernel='linear')

In [33]:
classifier.fit(x_train,y_train)

SVC(kernel='linear')

# model evaluation

#Accuracy Score

In [35]:
# Accuracy score on the training data 
X_train_prediction=classifier.predict(x_train)

In [36]:
training_data_accruacy= accuracy_score(X_train_prediction,y_train)

In [39]:
  print('Accuracy score of the training data : ', training_data_accruacy *100)

Accuracy score of the training data :  78.30342577487765


# From the traning data we can measure that model predicted 78% of accuracy

In [40]:
# now let us do it on Test data
# Accuracy score on the test  data 
X_test_prediction=classifier.predict(x_test)
test_data_accruacy= accuracy_score(X_test_prediction,y_test)

In [41]:
 print('Accuracy score of the test data : ', test_data_accruacy *100)

Accuracy score of the test data :  78.57142857142857


# Build  a predictive system to test the model

In [107]:
input_data=(9,171,110,24,240,45.4,0.721,54)

In [108]:
# Changing the input data to numpy array
input_data_as_np_array= np.asarray(input_data)

In [109]:
input_data_reshaped=input_data_as_np_array.reshape(1,-1) # tells the model to predict the label for only one array instead of completed dataset hence 1,-1

In [110]:
# again we have to scale it as we ahve taken the example without standardizing 

In [111]:
std_data=scaler.transform(input_data_reshaped)

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  "X does not have valid feature names, but"


In [112]:
print(std_data)

[[1.5310923  1.56897841 2.11300421 0.21841911 1.38963075 1.70086362
  0.75261125 1.76940112]]


In [113]:
prediction= classifier.predict(std_data)

In [114]:
if (prediction==0):
  print(" The Person is  Not diabetic as the ladel is",prediction)
else:
  print (" The person is a diabetic as the outcome is ",prediction)

 The person is a diabetic as the outcome is  [1]
